### DISCUSSION TITLE
Combine `matchPackageNames` and `matchSourceUrlPrefixes` with OR relation?

### DISCUSSION THREAD

**Discussion by TWiStErRob** [renovatebot/renovate#25580] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

github.com

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Wanted end result.

One PR updating both
 * `io.gitlab.arturbosch.detekt:detekt-gradle-plugin`
 * `io.gitlab.arturbosch.detekt`

### What you tried so far.

https://github.com/TWiStErRob/renovate-detekt-test (archived as https://github.com/TWiStErRob/repros/commit/4c22b0e7ccc4ffa43e4509562ccf0d56583b03d1 in case the repo is deleted)

I've managed to isolate a minimal repro as:
```json
{
	"groupName": "Detekt monorepo",
	"description": "Group together all dependencies of Detekt.",
	"matchSourceUrlPrefixes": [
		"https://github.com/detekt/detekt"
	],
	"matchPackageNames": [
		"io.gitlab.arturbosch.detekt",
		"io.gitlab.arturbosch.detekt:detekt-gradle-plugin",
		"io.gitlab.arturbosch.detekt:io.gitlab.arturbosch.detekt.gradle.plugin"
	]
}
```

when I remove `matchSourceUrlPrefixes` I get the result I want, but I want to keep that for other (unknown, changing) artifacts, e.g. `io.gitlab.arturbosch.detekt:detekt-cli`.

In my understanding `match...` should be "OR"'d together, although I couldn't find the relevant docs.

### Relevant debug logs

<details><summary>Logs</summary>

```json
 packageFiles with updates
{
  "baseBranch": "main"
  "config": {
    "gradle": [
      {
        "datasource": "maven",
        "deps": [
          {
            "currentValue": "1.23.1",
            "currentVersion": "1.23.1",
            "datasource": "maven",
            "depName": "io.gitlab.arturbosch.detekt:detekt-gradle-plugin",
            "depType": "dependencies",
            "fileReplacePosition": 21,
            "fixedVersion": "1.23.1",
            "groupName": "detekt",
            "homepage": "https://detekt.dev",
            "isSingleVersion": true,
            "managerData": {
              "fileReplacePosition": 21,
              "packageFile": "gradle/libs.versions.toml"
            },
            "packageName": "io.gitlab.arturbosch.detekt:detekt-gradle-plugin",
            "registryUrl": "https://repo.maven.apache.org/maven2",
            "registryUrls": [],
            "sourceUrl": "https://github.com/detekt/detekt",
            "versioning": "gradle",
            "warnings": [],
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "1.23.3",
                "newValue": "1.23.3",
                "releaseTimestamp": "2023-10-31T15:38:15.000Z",
                "newMajor": 1,
                "newMinor": 23,
                "updateType": "patch",
                "branchName": "renovate/detekt-monorepo"
              }
            ]
          },
          {
            "currentValue": "1.23.1",
            "currentVersion": "1.23.1",
            "datasource": "maven",
            "depName": "io.gitlab.arturbosch.detekt",
            "depType": "plugin",
            "fileReplacePosition": 21,
            "fixedVersion": "1.23.1",
            "groupName": "detekt",
            "isSingleVersion": true,
            "managerData": {
              "fileReplacePosition": 21,
              "packageFile": "gradle/libs.versions.toml"
            },
            "packageName": "io.gitlab.arturbosch.detekt:io.gitlab.arturbosch.detekt.gradle.plugin",
            "registryUrl": "https://plugins.gradle.org/m2",
            "registryUrls": [
              "https://plugins.gradle.org/m2/"
            ],
            "versioning": "gradle",
            "warnings": [],
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "1.23.3",
                "newValue": "1.23.3",
                "releaseTimestamp": "2023-11-04T09:07:26.000Z",
                "newMajor": 1,
                "newMinor": 23,
                "updateType": "patch",
                "branchName": "renovate/detekt"
              }
            ]
          }
        ],
        "packageFile": "gradle/libs.versions.toml"
      }
    ]
  }
}
```

</details>

**Comment by rarkins** [comment, marked as answer]
Only matchDepX and matchPackageX fields are OR'd together. Everything else is AND. You need two package rules

**Reply by TWiStErRob** [reply]
Ah! So that's what the "either" word is documenting at https://docs.renovatebot.com/configuration-options/#packagerules. The "everything else is AND" is implied.

I did it and the result is:
https://github.com/TWiStErRob/renovate-detekt-test/pull/7
Thank you!

I had a question: If I have two separate rules, can they officially have the same groupName to simulate OR? And the answer is yes, based on the above PR.

I wonder if there's a way to automatically detect these "you're probably shooting yourself in the foot" configurations? Do you know of a static analyser that does anything like this (other than the maintainers' brains)?

**Reply by rarkins** [reply]
Looks like we should make the OR/AND clearer in docs.

It's totally ok to have same groupName in different rules 

We have some validation checks to catch *definite* problems in packageRules but does not "warn" like you're hoping for

**Reply by TWiStErRob** [reply]
Tried to document this here: https://github.com/renovatebot/renovate/pull/29264
